OemPkg/SecureBootRecovery/SecureBootRecovery.c

```diff
@@ -0,0 +1,136 @@
+  @retval EFI_INVALID_PARAMETER SystemTable provided was not valid.
+  @retval other                 Some error occurs when executing this entry point.
+
+**/
+EFI_STATUS
+EFIAPI
+UefiMain (
+  IN EFI_HANDLE        ImageHandle,
+  IN EFI_SYSTEM_TABLE  *SystemTable
+  )
+{
+  EFI_STATUS  Status;
+  UINT32      Attributes;
+
+  Attributes = EFI_VARIABLE_NON_VOLATILE |
+               EFI_VARIABLE_BOOTSERVICE_ACCESS |
+               EFI_VARIABLE_RUNTIME_ACCESS |
+               EFI_VARIABLE_TIME_BASED_AUTHENTICATED_WRITE_ACCESS |
+               EFI_VARIABLE_APPEND_WRITE;
+
+  //
+  // Start checking that the system is in a state we can safely use
+  //
+  if (SystemTable == NULL) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  if ((SystemTable->ConOut == NULL) || (SystemTable->ConOut->OutputString == NULL) || (SystemTable->ConOut->ClearScreen == NULL)) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  if ((SystemTable->BootServices == NULL) || (SystemTable->BootServices->Stall == NULL)) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  //
+  // After this point, we should be able to print and stall but nothing else has been verified
+  //
+  if ((SystemTable->RuntimeServices == NULL) || (SystemTable->RuntimeServices->ResetSystem == NULL)) {
+    Status = EFI_INVALID_PARAMETER;
+    goto Exit;
+  }
+
+  //
+  // Start informing the user of what is happening
+  //
+  Status = SystemTable->ConOut->ClearScreen (SystemTable->ConOut);
+  if (EFI_ERROR (Status)) {
+    goto Exit;
+  }
+
+  Status = SystemTable->ConOut->OutputString (SystemTable->ConOut, L"\r\nAttempting to update the system's secureboot certificates\r\n");
+  if (EFI_ERROR (Status)) {
+    goto Exit;
+  }
+
+  SystemTable->ConOut->OutputString (SystemTable->ConOut, L"Learn more about this tool at https://aka.ms/securebootrecovery\r\n");
+
+  //
+  // Perform the append operation
+  //
+  Status = SystemTable->RuntimeServices->SetVariable (
+                                           L"db",
+                                           &gEfiImageSecurityDatabaseGuid,
+                                           Attributes,
+                                           sizeof (mDbUpdate),
+                                           mDbUpdate
+                                           );
+  if (EFI_ERROR (Status)) {
+    //
+    // On failure, inform the user and reboot
+    // Likely this will continue to fail on reboot, the user will hopefully go to https://aka.ms/securebootrecovery to learn more
+    //
+    SystemTable->ConOut->OutputString (SystemTable->ConOut, L"\r\nFailed to update the system's secureboot keys\r\n");
+    goto Reboot;
+  }
+
+  //
+  // Otherwise the system took the update, so let's inform the user
+  //
+  SystemTable->ConOut->OutputString (SystemTable->ConOut, L"\r\nSuccessfully updated the system's secureboot keys\r\n");
+
+Reboot:
+
+  //
+  //  Stall for 10 seconds to give the user a chance to read the error message

```

Comment:
Technically, just read the message since the successful case will drop to here as well.

```suggestion
  //  Stall for 10 seconds to give the user a chance to read the message

```